src/Compiler/Microsoft.AspNetCore.Razor.Language/src/CodeGeneration/CodeWriter.cs

```diff
@@ -142,95 +252,98 @@ public CodeWriter Write(string value, int startIndex, int count)
         {
-            // This is newline that's spread across two writes. Skip the first character of the
-            // current write operation.
-            //
-            // We don't need to increment our newline counter because we already did that when we
-            // saw the \r.
-            i += 1;
-            trailingPartStart = 1;
+            // Skip the first character of span to ensure that it isn't considered in the following
+            // line break detection loop.
+            span = span[1..];
         }
 
-        // Iterate the string, stopping at each occurrence of a newline character. This lets us count the
-        // newline occurrences and keep the index of the last one.
-        var endIndex = startIndex + count;
-        while (i < endIndex)
+        // Iterate the span, stopping at each occurrence of a new-line character.
+        // This let's us count the new-line occurrences and keep the index of the last one.
+        int newLineIndex;
+        while ((newLineIndex = span.IndexOfAny('\r', '\n')) >= 0)
         {
-            var ch = value[i];
+            _currentLineIndex++;
+            _currentLineCharacterIndex = 0;
 
-            i++;
+            newLineIndex++;
 
-            if (ch is '\r' or '\n')
+            // We might have stopped at a \r, so check if it's followed by \n and then advance the index.
+            // Otherwise, we'll count this line break twice.
+            if (newLineIndex < span.Length &&
+                span[newLineIndex - 1] == '\r' &&
+                span[newLineIndex] == '\n')
             {
-                // Newline found.
-                _currentLineIndex++;
-                _currentLineCharacterIndex = 0;
-
-                // We might have stopped at a \r, so check if it's followed by \n and then advance the index.
-                // Otherwise, we'll count this line break twice.
-                if (ch == '\r' &&
-                    i < endIndex &&
-                    value[i] == '\n')
-                {
-                    i++;
-                }
-
-                // The 'suffix' of the current line starts after this newline token.
-                trailingPartStart = i;
+                newLineIndex++;
             }
-        }
 
-        if (trailingPartStart == null)
-        {
-            // No newlines, just add the length of the data buffer
-            _currentLineCharacterIndex += count;
-        }
-        else
-        {
-            // Newlines found, add the trailing part of 'data'
-            _currentLineCharacterIndex += endIndex - trailingPartStart.Value;
+            span = span[newLineIndex..];
         }
 
+        _currentLineCharacterIndex += span.Length;
+
         return this;
     }
 
     public CodeWriter WriteLine()
     {
-        _builder.Append(NewLine);
-
-        _currentLineIndex++;
-        _currentLineCharacterIndex = 0;
-        _absoluteIndex += NewLine.Length;
+        WriteCore(_newLine, allowIndent: false);
```

Comment:
It seems that previously this was implemented to skip the complex algorithm determining `_currentLineIndex` since we know this won't be a partial line (now we could do the same via `AddTextChunk` + manual update of the fields as previously). But perhaps that's a useless optimization.

Reply:
Yeah, I don't think the optimization is all that useful and wanted to route everything through the same logic to update the `_currentLineIndex` and `_currentLineCharacterIndex`, since there were bugs in the original implementation.